### PR TITLE
JAI-173: Replaced the HTTP Client

### PR DESCRIPTION
**Pull request by martinmurtiono**

### CODE REVIEW

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 1]
```diff
@@ -3,173 +3,94 @@
...
-                    }
-                }
-            }
-            return objectMapper.readTree(apiResults);
-        } catch (final IOException e) {
-            final StringBuilder sb = new StringBuilder();
-
-            sb.append("{");
-            sb.append("\"Request\": {");
+        final OkHttpClient newClient = client.newBuilder().build();
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Why are we building a new client here? Why not just use the existing one instantiated in the constructor?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 2]
```diff
@@ -3,173 +3,94 @@
...
+        final Request request = new Request.Builder()
+                .header("Authorization", this.authentication)
+                .header("Cache-Control", "no-store")
+                .url(url)
+                .build();
+        final Response response = newClient.newCall(request).execute();
+        try (ResponseBody responseBody = response.body()){
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Nitpick: final throughout

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 3]
```diff
@@ -3,173 +3,94 @@
...
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-                sb.append("\"Code\": ").append(urlConnection.getResponseCode()).append(",");
-                sb.append("\"Message\": \"").append(urlConnection.getResponseMessage()).append("\",");
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(responseBody.string()).append("\"");
+                sb.append("}");
+                sb.append("}");
+                log.debug("Encountered connection error: " + sb);
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Why is this at DEBUG level and not at ERR?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 4]
```diff
@@ -3,173 +3,94 @@
...
+                throw new IOException(String.valueOf(response));
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Generally you should either log a message or throw, but not both. If you do both, you end up with double logging (because whomever catches this exception will presumably also log an error too).

(Yes, I see the existing code logs and throws as well; it's still worth pointing out.)

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 5]
```diff
@@ -3,173 +3,94 @@
...
     private final String authentication;
-    private String jsessionId = null;
-    private String upstream = null;
-    private String cookies = "";
-    private String pinnedNode = null;
 
     public ApiCaller(final JiraActionsIndexBuilderConfig config) {
         this.config = config;
         this.authentication = getBasicAuth();
+        this.client = new OkHttpClient.Builder()
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: By default OkHTTP [rejects all cookies](https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.Builder.html#cookieJar-okhttp3.CookieJar-). You'll need to provide a CookieJar.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 6]
```diff
@@ -3,173 +3,94 @@
...
+                .connectionPool(new ConnectionPool())
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: `new ConnectionPool())` is already the default, no need to specify.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 7]
```diff
@@ -3,173 +3,94 @@
...
+                .connectTimeout(2, TimeUnit.MINUTES)
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Where did these values come from?

Also, putting one in MINUTES and one in MILLISECONDS is weird, especially when the one in MILLISECONDS could be specified in seconds.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 8]
```diff
@@ -3,173 +3,94 @@
...
+                        value = String.join(",", responseHeaders.value(i));
                     }
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-                sb.append("\"Code\": ").append(urlConnection.getResponseCode()).append(",");
-                sb.append("\"Message\": \"").append(urlConnection.getResponseMessage()).append("\",");
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(responseBody.string()).append("\"");
+                sb.append("}");
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Nit: Can replace `sb.append("}");` with `sb.append('}');` There's an IntelliJ inspection you can turn on to report these things; it claims there's a minor performance gain from not having to turn the character into a string.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 9]
```diff
@@ -3,178 +3,120 @@
...
-    private String jsessionId = null;
-    private String upstream = null;
-    private String cookies = "";
-    private String pinnedNode = null;
 
     public ApiCaller(final JiraActionsIndexBuilderConfig config) {
         this.config = config;
         this.authentication = getBasicAuth();
+        this.client = new OkHttpClient.Builder()
+                .connectTimeout(20, TimeUnit.SECONDS)
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Where did these numbers come from?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 10]
```diff
@@ -3,178 +3,120 @@
...
         final String basicAuth = "Basic " + new String(new Base64().encode(userPass.getBytes()));
         return basicAuth;
     }
+
+    private static CookieJar setupCookieJar() {
+        return new CookieJar() {
+            private final HashMap<String, List<Cookie>> cookieStore = new HashMap<>();
+
+            @Override
+            public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Does this work? It looks like `HttpUrl` contains path segments, which means every URL will be unique. One hack would be to simply use a single List of cookies, since we know we're only connecting to a single domain and all the cookies are safe to share.

Also consider something like this: https://github.com/franmontiel/PersistentCookieJar or other alternatives in this thread: https://github.com/square/okhttp/issues/2890

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 11]
```diff
@@ -3,178 +3,120 @@
...
+                cookieStore.put(url.host(), cookies);
+            }
+
+            @Override
+            public List<Cookie> loadForRequest(HttpUrl url) {
+                List<Cookie> cookies = cookieStore.get(url.host());
+                return cookies != null ? cookies : new ArrayList<Cookie>();
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Java 1.8 actually introduces a really cool function to do this, `computeIfAbsent()`. Something like `cookies.computeIfAbsent(cookies, () -> new ArrayList<Cookie>());` Prefer `computeIfAbsent()` to `getOrDefault()` because `cookies.getOrDefault(cookies, new ArrayList<Cookie());` would instantiate a new `ArrayList` every time, even if the map contains the value. The `Supplier<>` provides extra indirection that prevents this in `computeIfAbsent()`.

(I guess technically it would be even slightly more efficient to refactor the Supplier into a private variable.)

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 9 above)
Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: They default to 10 seconds. After messing around with the batch size to around 100 I set it around there because towards the end of indexing all the issues, the time could spike to 15-20 seconds. This shouldn't be a problem now that we are indexing smaller amounts and I can remove it (set it back to default).

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 10 above)
Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Tried importing the PersistentCookieJar, but it was in a different repository and was acting up. I'm gonna try going with the single List of cookies approach.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 9 above)
Comment (Anusien) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Feels like the kind of thing that should be sourced from a config file.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 10 above)
Comment (Anusien) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Looks like PersistentCookieJar is too tightly coupled on Android stuff for us anyway.

Depending on how OkHttp uses the CookieJar, we may want a Set<> instead of a List<>.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 12]
```diff
@@ -3,178 +3,108 @@
...
+        final String anodeId = response.header("X-ANODEID");
+
+        if(!Objects.equals(pinnedNode, anodeId)) {
+            if (pinnedNode != null) {
+                log.warn("Expected X-ANODEID={} but found {}", pinnedNode, anodeId);
             }
-            return objectMapper.readTree(apiResults);
-        } catch (final IOException e) {
-            final StringBuilder sb = new StringBuilder();
+            cookies = String.join(";", response.headers("Set-Cookie"));
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: We wouldn't want to replace all the cookies we have with whatever is new, because that would be throwing away old cookies.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 13]
```diff
@@ -3,178 +3,108 @@
...
+                .header("Authorization", this.authentication)
+                .header("Cache-Control", "no-store")
+                .header("Cookie", cookies)
+                .url(url)
+                .build();
+        final Response response = client.newCall(request).execute();
+
+        final String anodeId = response.header("X-ANODEID");
+
+        if(!Objects.equals(pinnedNode, anodeId)) {
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: I think we want to do this always, not just when there's a mismatch. If JIRA wants us to set a cookie, we probably should.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 14]
```diff
@@ -3,178 +3,108 @@
...
-                        setCookies();
-                        pinnedNode = anodeId;
-                        log.info("Set JSESSION={};upstream={}. Pinning to X-ANODEID={}",
-                                jsessionId, upstream, anodeId);
-                    }
-                }
+        final Request request = new Request.Builder()
+                .header("Authorization", this.authentication)
+                .header("Cache-Control", "no-store")
+                .header("Cookie", cookies)
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Manually handling the cookies seems way more cumbersome than just implementing a CookieJar

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 15]
```diff
@@ -3,178 +3,130 @@
...
+import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 import java.util.Objects;
 
 public class ApiCaller {
     protected final JiraActionsIndexBuilderConfig config;
 
-    private static final Logger log = LoggerFactory.getLogger(ApiCaller.class);
+    private static final Logger log = LoggerFactory.getLogger(IssuesAPICaller.class);
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Why did you change this?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 16]
```diff
@@ -3,178 +3,130 @@
...
         return basicAuth;
     }
+
+    private static CookieJar setupCookieJar() {
+        return new CookieJar() {
+            private List<Cookie> cookies = new ArrayList<>();
+
+            @Override
+            public void saveFromResponse(final HttpUrl url, final List<Cookie> cookies) {
+                if(!cookies.equals(this.cookies)) {
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: It's not correct to just replace the current set of cookies with the incoming list. You'll end up throwing away cookies that were correctly set.

https://github.com/indeedeng/imhotep-builder-jira/pull/21#discussion_r311771639

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 17]
```diff
@@ -49,6 +50,7 @@ public JsonNode getJsonNode(final String url) throws IOException {
                 .build();
         final Response response = client.newCall(request).execute();
 
+        log.debug(""+cookieJar.loadForRequest(HttpUrl.parse(url)));
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Did you mean to leave this in?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 18]
```diff
@@ -3,178 +3,130 @@
...
+        if(!Objects.equals(pinnedNode, anodeId)) {
+            if (pinnedNode != null) {
+                log.warn("Expected X-ANODEID={} but found {}", pinnedNode, anodeId);
             }
-            return objectMapper.readTree(apiResults);
-        } catch (final IOException e) {
-            final StringBuilder sb = new StringBuilder();
+            pinnedNode = anodeId;
+        }
+        try (final ResponseBody responseBody = response.body()){
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Nit: spacing between `)` and `{`

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 19]
```diff
@@ -3,178 +3,130 @@
...
+                sb.append('}');
+                sb.append('}');
+                throw new IOException(String.valueOf(response));
             }
-
-            sb.append("}");
-            sb.append("}");
-            log.error("Encountered connection error: " + sb);
+            return objectMapper.readTree(responseBody.string());
+        } catch (IOException e) {
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: Why are we catching this exception only to rethrow it?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 20]
```diff
@@ -3,178 +3,130 @@
...
-import java.util.Map;
 import java.util.Objects;
 
 public class ApiCaller {
     protected final JiraActionsIndexBuilderConfig config;
 
-    private static final Logger log = LoggerFactory.getLogger(ApiCaller.class);
+    private static final Logger log = LoggerFactory.getLogger(IssuesAPICaller.class);
     private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final CookieJar cookieJar = setupCookieJar();
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: I'm not sure it matters, but why is this static?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 17 above)
Comment (Anusien) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: What does it even mean to call `toString()` on this List? Presumably you'd want some sort of `String.join` or something instead.

Also, prefer explicit `.toString()` rather than getting it by default by concatenating with a string.

Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: It was just a quick debug to make sure it was working, didn't mean to leave it in.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 20 above)
Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: I originally had CookieJar as a static class and just left it in. I guess we can just call setupCookieJar() in the client builder now.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 19 above)
Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: I modeled this kind of similar to the existing code so it logs the stringbuilder. Should we just not do that and log the error instead?

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java (hunk 20 above)
Comment (Anusien) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: CookieJar as a static class makes sense. But being a static variable means that if there were to be multiple instances of the ApiCaller class, they would share a single CookieJar. This isn't a good idea. Note that the CookieJar you've built isn't thread-safe, so if we were ever to make calls in parallel, that would be a problem.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 21]
```diff
@@ -22,159 +28,107 @@
...
         return basicAuth;
     }
+
+    private static CookieJar setupCookieJar() {
+        return new CookieJar() {
+            private final Map<String, Cookie> cookies = new HashMap<>();
+            @Override
+            public void saveFromResponse(final HttpUrl url, final List<Cookie> cookies) {
+                cookies.forEach(cookie -> {
+                    this.cookies.remove(cookie.name());
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: I know I suggested this, but looking at it again, I think the remove() call is redundant.

File: src/main/java/com/indeed/jiraactions/api/ApiCaller.java [hunk 22]
```diff
@@ -22,159 +28,102 @@
...
-                try {
-                    br.close();
-                } catch (final IOException ignored) {
-                }
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(response.body().string()).append("\"");
+                sb.append('}');
+                sb.append('}');
+                log.error("Encountered connection error: " + sb);
```
Comment (Anusien) [inline src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: So let's say an error occurs. We log the error, then what happens?

Comment (Martin-Murtiono) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: It logs the request attempt and then throws the IOException

Comment (Anusien) [reply src/main/java/com/indeed/jiraactions/api/ApiCaller.java]: If the HTTP request completes without throwing an exception but `response.isSuccessful()` returns false (meaning it has a 4xx or 5xx response code), your code will log the error and then attempt to deserialize the responseBody and return that. I suspect that will cause an exception, but I'm not positive.

If the HTTP request completes exceptionally (because of connectivity problems or timeouts), your code will not log anything here at all and rely on the caller catching the exception.


There's some nuance in the old connection code you may be missing; `HttpURLConnection.getInputStream()` would throw an `IOException` if the server you were talking to returned a 4xx or a 5xx code.
